### PR TITLE
Added the function isPromotion 

### PR DESCRIPTION
**Pull request by martinreycristina**
We have made the function isPromotion: Checks if the move resulting from a source/target is a pawn promotion. It takes in consideration:
 - ... If the move is valid.
 - ... If the piece is a pawn.
 - ... And if the rank the pawn moves to is the correct one for a promotion.

Nonetheless, we made a battery of tests that validates the functionality of the feature.

This pull request is related to the following issue: https://github.com/pheyvaer/solid-chess/issues/55

### CODE REVIEW

**Comment by pheyvaer** [comment]
Thanks for the PR! Will have a look ASAP.

**Comment by pheyvaer** [comment]
I have a few questions:

1) What about the original `promotion` field provided by chess.js? Does this already do some validation of a validate promotion?
2) I looks weird that you first do the move just to check if its valid and then undo it. Is this really needed? And even more, what is the expected use of this method, i.e., where in the flow of calls does it happen?
3) Where do you specify the actual promotion? You say that the player should have the option to choose, but I don't see that reflected in the code.

**Comment by srensamblador** [comment]
Hi,
* There's no built in way to check for promotion yet, AFAIK. Therefore we opted to do the checks ourselves. However, there's work towards a new API call to check for promotions: https://github.com/jhlywa/chess.js/pull/148 
* There's another way working with the public API, which would be generating a list of valid moves from the source square and checking if the move we want to do is in there. We'd be happy to change it if you prefer this (in hindsight we do prefer it). 
    The need for the valid move checking is to prevent the GUI / CLI from asking the player for a piece choice for the promotion when the move wasnt valid in the first place. For example maybe the move meets the promotion requirements but cant be performed because the king is in check.
* Our plan was to provide is_promotion just as a check in semantic-chess and then on solid-chess' on_drop() (in GUI's case for example) this function would be called and if it truly is a promotion the chess implementation would show a popup for instance, that would allow the player to choose among the pieces you can promote into.

Let us know any change/feature you want us to do :)

**Comment by pheyvaer** [comment]
1) That PR is open for quite some time. You could try to ping them to see if something is moving with the PR, but I can understand you don't get a response quick or that the PR is accepted quickly.
2) OK, well, both work and I can live with the undo, considering (1). But the isPromotion method does only check for promotion: when the promotion is valid the move is also done. The move should be undo in both cases.
3) :+1: 
4) Is the promotion part of the same move or is the promotion part a separate move. Following that, how is this done in the corresponding RDF?

**Comment by srensamblador** [comment]
1. I'll ping them ASAP
2. I'm not sure I understand what you mean. If the move is invalid it's not performed and if its valid it's undoed.
4. We don't think isPromotion should interact with the RDF since the move will be performed in semantic-chess' doMove (with the selected piece for promotion already sent in the move after the user chooses one). We think this issue is best kept isolated from the RDF side as it's more on the chess.js one from our POV.
 Let us know what you think

**Comment by pheyvaer** [comment]
1) :+1: 
2) You're right. Sorry, my bad.
3) It's probably directly encoded in the SAN, right? In that case indeed no changes to the RDF are needed :smile:

**Comment by martinreycristina** [comment]
Yes, the notation that we use to pass the parameters is SAN.  We don't need to touch anything from the RDF side.
If you need anything else, don't hesitate to tell us.
Is this ok for you? :)

File: lib/semanticchess.js [hunk 1]
```diff
@@ -191,6 +191,39 @@ class SemanticChess {
...
+    if (move === null)
+      return false;
+    else
+      this.chess.undo();
+    
+    var rankSource = source.substring(2,1);
+    var rankTarget = target.substring(2,1);
+    var piece = this.chess.get(source).type;
+
+    if (piece === 'p' &&
```
Comment (pheyvaer) [inline lib/semanticchess.js]: This `if` together with the `return false` could all be turned into one `return`.

File: lib/semanticchess.js [hunk 2]
```diff
@@ -191,6 +191,39 @@ class SemanticChess {
...
+   * This method evaluates a move to check if it is a pawn promotion
+   * For that it must:
+   * - Be valid
+   * - Be a pawn
+   * - Move onto rank 8 or 1 from the previous ones
+   * @param {string} source Position of the move's start 
+   * @param {string} target Position of the move's destination
+   */
+  isPromotion(source, target){
+    var moveToTest = {
```
Comment (pheyvaer) [inline lib/semanticchess.js]: Use `const` were possible.

**Comment by pheyvaer** [comment]
OK, I added two comments about the code itself, if you update that. I will check if everything builds locally and if the tests succeed.

**Comment by martinreycristina** [comment]
We have already done the changes you commented. 
The tests we have created give the go-ahead to the new code.
Anyway, you can pass your tests to see if it works the way you wanted.
Thank you for commenting on that and trusting us.

File: lib/semanticchess.test.js [hunk 3]
```diff
@@ -172,7 +173,68 @@ describe('Semantic Chess', function() {
...
+		  moveBaseUrl: 'http://example.org/myfile',
+		  userWebId: 'http://example.org/#me',
+		  opponentWebId: 'http://example.org/#other',
+		  name: 'My game',
+		  uniqid: Utils.customUniqid()
+		});
+		
+		//NEGATIVE TESTS
+		//If the king is in check
+		semanticChess.chess = new chess.Chess('8/4PP1k/8/7r/q7/8/K7/2n4Q w - - 0 1');
```
Comment (pheyvaer) [inline lib/semanticchess.test.js]: Can you change this to use the Semantic Chess constructor? I know you can directly access the `chess` variable of an instance, but it is cleaner to use the constructor. Also if users check the test code for usage example they might get the wrong idea about how to use the library.

**Comment by pheyvaer** [comment]
Sorry for the delay, regarding the test code. If that is fixed I'm happy to merge :)

**Comment by martinreycristina** [comment]
Hi @pheyvaer . We've already done what you asked us to do, but take a look at it to confirm that's what you wanted.

File: lib/semanticchess.test.js [hunk 4]
```diff
@@ -178,60 +177,98 @@ describe('Semantic Chess', function() {
...
-		});
 		
 		//POSITIVE TESTS
+		
 		//Test in the middle of the board (white pawn)
-		semanticChess.chess = new chess.Chess('8/4P3/8/8/8/8/K6k/8 w - - 0 1');
+		var semanticChess = new SemanticChess({
+		  params,
+		  startPosition: '8/4P3/8/8/8/8/K6k/8 w - - 0 1'
+		});
```
Comment (pheyvaer) [inline lib/semanticchess.test.js]: The constructor is wrong here. To achieve what you want to do you should do `params.startPosition= "..."` and then do `new SemanticChess(params)`.

**Comment by pheyvaer** [comment]
Add one last remark. Almost there! 😉

**Comment by martinreycristina** [comment]
Take a look at the changes we've made this time, the code is getting the right way!

**Pull request merged**